### PR TITLE
OBJ-231 Integrate action code validation into create objection

### PR DESCRIPTION
**Pull request by markpit**
Plugging in the actionCodeValidator into ObjectionService so we can validate the action code when creating a new Objection

Also, the ObjectionService constructor args list had grown too large (Sonar) so had to move them out into fields.

### CODE REVIEW

**Review by ICoxCH — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java [hunk 1]
```diff
@@ -86,7 +86,7 @@ public Objection build() {
     @Field("status")
     private ObjectionStatus status;
     @Field("action_code")
-    private int actionCode;
+    private Long actionCode;
```
Comment (mattch1) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: Should this be primitive in the object itself?

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: shouldn't matter due to auto boxing, just that i'd done the validation with Long so tried to keep it consistent.

File: src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java [hunk 2]
```diff
@@ -95,37 +102,78 @@
...
         assertEquals(MOCKED_TIME_STAMP, objectionResponse.getCreatedOn());
         assertEquals(COMPANY_NUMBER, objectionResponse.getCompanyNumber());
-        assertEquals(ObjectionStatus.OPEN, objectionResponse.getStatus());
+        assertEquals(OPEN, objectionResponse.getStatus());
         assertEquals(AUTH_ID, objectionResponse.getCreatedBy().getId());
         assertEquals(E_MAIL, objectionResponse.getCreatedBy().getEmail());
     }
 
+    @Test
+    void createObjectionIneligibleStatusTest() throws ValidationException {
```
Comment (mattch1) [inline src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java:154]: Is there a test scenario for 209 ticket where the code is neither eligible nor ineligible - is that covered anywhere on another ticket perhaps?

File: src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java [hunk 3]
```diff
@@ -69,6 +74,8 @@
     private static final String ATTACHMENT_ID = "12348765";
     private static final String REASON = "REASON";
     private static final String ACCESS_URL = "/dummyUrl";
+    private static final Long ACTION_CODE_OK = 3000L;
```
Comment (mattch1) [inline src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java:77]: Are these made up action/ineligibility codes for test purposes?

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java (hunk 1 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: As long as there's no danger of Long being null that should be fine.

File: src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java (hunk 3 above)
Comment (markpit) [reply src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java:77]: yes, it's just to test that the same value is being populated in the objection that is saved to mongo.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java (hunk 1 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: I hadn't thought of a null situation. I'll change to primitive then.

File: src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java (hunk 2 above)
Comment (markpit) [reply src/test/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionServiceTest.java:154]: The AllowedValuesValidationRuleTest should cover that scenario, which basically checks that if a value is in the list of 'ok' values, it is ok, if it is not in the list then a validation exception gets thrown.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java (hunk 1 above)
Comment (ICoxCH) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: What would be better a null pointer exception or a default value of 0 being assigned?

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: actually, yes, a null in mongo would indicate a missing action code, perhaps companies can have no action code anyway? 0 might indicate an actual 0 code or possible missing code. I think i'll leave it as is then.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/validation/AllowedValuesValidationRule.java [hunk 4]
```diff
@@ -22,7 +22,7 @@ public AllowedValuesValidationRule(List<T> allowableValues,
     @Override
     public void validate(T input, String logContext) throws ValidationException {
         if (!allowableValues.contains(input)) {
-            apiLogger.debugContext(logContext, String.format("%s is not an allowed value", input));
+            apiLogger.debugContext(logContext, String.format("AllowedValuesValidationRule %s is not an allowed value", input));
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/validation/AllowedValuesValidationRule.java:25]: You could use Class.getSimpleName() here instead of repeating the class name. Same in the other rule log statement.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java (hunk 1 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: We could have -1 - just think of what we return to the web module in this scenario.

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: I guess you could still have a -1 action code in chips DB

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: As long the web will handle a value of null we should be ok.

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: thinking about it, we don't actually return the action code to the web, just the objection status (enum value)

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java [hunk 5]
```diff
@@ -94,12 +93,36 @@ public Objection createObjection(String requestId, String companyNumber, String
...
+
+        logger.debugContext(requestId, "Company action code is " + actionCode);
+
+        return actionCode;
+    }
+
+    private ObjectionStatus getObjectionStatusForCreate(Long actionCode, String companyNumber, String logContext) {
+        ObjectionStatus objectionStatus = ObjectionStatus.OPEN;
+        try {
+            actionCodeValidator.validate(actionCode, logContext);
```
Comment (mattch1) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java:114]: Might be a good idea to have a null check here if the action code is to be Long.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java (hunk 1 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/entity/Objection.java:89]: of course this action code does not get returned to the web - think of the wrong place. However, we may want to check for null where this action code intersects with creating the status that we do return to the web - added another comment below.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java (hunk 5 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java:114]: Maybe I shouldn't have used a non-primitive type in my earlier PR to get the action code, thereby making it clear (enforcing) that non-null values are just not allowed?

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java:114]: > Might be a good idea to have a null check here if the action code is to be Long.

if it is null, it should be validated as normal. The AllowedValuesValidationRule will throw a ValidationException containing ObjectionStatus.INELIGIBLE_NO_DISSOLUTION_ACTION if it is null, as null is not one of the codes specified in OBJ-209 (these codes are in config ACTION_CODES_STRIKE_OFF_NOTICE). I can add an extra unit test for that rule to test a null value being passed into it.

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java:114]: > Maybe I shouldn't have used a non-primitive type in my earlier PR to get the action code, thereby making it clear (enforcing) that non-null values are just not allowed?

I think it's ok, it covers a scenario where a company has no action code

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/validation/AllowedValuesValidationRule.java (hunk 4 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/validation/AllowedValuesValidationRule.java]: updated to use getSimpleName()

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java (hunk 5 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java:114]: I've added extra unit tests to the validation rules and they work as expected when passing null into them.

**Review by mattch1 — APPROVED** [review]
Good to have those extra unit tests.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java (hunk 5 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/service/impl/ObjectionService.java:114]: > > Maybe I shouldn't have used a non-primitive type in my earlier PR to get the action code, thereby making it clear (enforcing) that non-null values are just not allowed?
> 
> I think it's ok, it covers a scenario where a company has no action code

Oh yes, true. And just seen that it is a nullable column in the database (although I did think that a company always had an action code).

**Review by mwestacott — APPROVED** [review]
